### DISCUSSION TITLE
Why do we need `add_self_loops`?

### DISCUSSION THREAD

**Discussion by quy-ng** [pyg-team/pytorch_geometric#3285] [Q&A] [answered]
Hi team, I knew there is a function named `add_self_loops`, but I struggle to understand why we should use it?

In the document, it is described  _Adds a self-loop  to every node  in the graph given by edge_index_ , but in [test case](https://github.com/pyg-team/pytorch_geometric/blob/74245f3a680c1f6fd1944623e47d9e677b43e827/test/utils/test_loop.py) occurs to me confusing.
`edge_index = torch.tensor([[0, 1, 0], [1, 0, 0]])`
`expected = [[0, 1, 0, 0, 1], [1, 0, 0, 0, 1]]`

Much appreciate if you could help me to understand

**Comment by wsad1** [comment, marked as answer]
So the `add_self_loops` function adds edges from each node to itself. Such a thing might be useful if you want to say pass a message from a node to itself , like lets say you have a graph with no self loops then a layer like `GATConv` will pay attention to only a nodes neighbors, but with [add_self_loops](https://github.com/pyg-team/pytorch_geometric/blob/2106bc741566395af02a4af0921162ce0245dc9e/torch_geometric/nn/conv/gat_conv.py#L156) it can pay some "attention" to itself too.

Now coming to the test case. `add_self_loops` doesn't check for existing self loops it simply appends all possible self loops to the `edge_index` so the expected output `expected = [[0, 1, 0, 0, 1], [1, 0, 0, 0, 1]]`, the first three edges `(0,1), (1,0), (0,0)` are already present and the self loops `(0,0), (1,1)`  are added the self loop `(0,0)` gets repeated. 

On another note you can check [add_remaining_self_loops](https://pytorch-geometric.readthedocs.io/en/latest/modules/utils.html#torch_geometric.utils.add_remaining_self_loops) which only adds self loops that are not already present.

**Reply by quy-ng** [reply]
thanks

**Reply by alexrutherford** [reply]
I wanted to ask if `add_self_loops` will solve the problem I have now. I basically want to train a GCN to fill in missing node level features. Currently, I have N node level features, and I want to be able to predict missing values in one of these features e.g. if a new node is added to the network (and I know which existing nodes it connects to).

So for training my features are the N node level features and the target is one of those N node level features. But obviously for the node I am interested in predicting, I only want to use the known feature values of the _neighbours_ and to somehow mask the known feature value of that node of interest. Otherwise I am just predicting X based on X + {others}!

Would the deletion of self loops in the last layer effectively do that? I am looking for examples of predicting missing node level attributes but I couldn't find any.

**Reply by wsad1** [reply]
`add_self_loops = False` will prevent a node from having itself as a neighbour.
But for your problem, setting this to `False` will help for layers like[ GCNConv](https://pytorch-geometric.readthedocs.io/en/latest/modules/nn.html#torch_geometric.nn.conv.GCNConv) where the node update doesn't use target node information, but it won't help for layers like [SAGEConv](https://pytorch-geometric.readthedocs.io/en/latest/modules/nn.html#torch_geometric.nn.conv.SAGEConv)  where node update depends on target node.

**Reply by rusty1s** [reply]
Please see my answer here https://github.com/pyg-team/pytorch_geometric/discussions/3428.
